File: src/models/Project.ts
import mongoose, { Schema, Document, Model } from 'mongoose';

export interface IProject extends Document {
  projectId: string;
  title: string;
  description: string;
  organizationId: mongoose.Types.ObjectId;
  createdBy: mongoose.Types.ObjectId;
  createdByRole?: string;
  visibleToAll?: boolean;
  techStack: string[];
  startDate: Date;
  endDate: Date;
  totalHours: number;
  linkedProjects?: mongoose.Types.ObjectId[];
  loggedHoursPerWeek: {
    weekStartDate: Date;
    totalHours: number;
  }[];
  status: 'planning' | 'active' | 'completed' | 'canceled' | 'on-hold';
  employeeContributions: {
    employeeId: string;
    userId?: mongoose.Types.ObjectId;
    role: string;
    isTeamLead: boolean;
    weeklyHours: {
      weekStartDate: Date;
      hours: number;
      reportedTech?: string[];
    }[];
  }[];
  department: string;
  priority: 'low' | 'medium' | 'high' | 'critical';
  risks: {
    description: string;
    severity: 'low' | 'medium' | 'high';
    status: 'identified' | 'mitigated' | 'resolved';
  }[];
  createdAt: Date;
  updatedAt: Date;
}

const ProjectSchema: Schema = new Schema({
  projectId: {
    type: String,
    required: [true, 'Project ID is required'],
    unique: true,
    index: true
  },
  title: {
    type: String,
    required: [true, 'Project title is required'],
    trim: true
  },
  description: {
    type: String,
    required: [true, 'Project description is required'],
    trim: true
  },
  organizationId: {
    type: Schema.Types.ObjectId,
    ref: 'Organization',
    required: true,
    index: true
  },
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  createdByRole: {
    type: String,
    index: true
  },
  visibleToAll: {
    type: Boolean,
    default: false,
    index: true
  },
  techStack: [{
    type: String,
    required: true
  }],
  startDate: {
    type: Date,
    required: true
  },
  endDate: {
    type: Date,
    required: true
  },
  totalHours: {
    type: Number,
    default: 0
  },
  linkedProjects: [{
    type: Schema.Types.ObjectId,
    ref: 'Project'
  }],
  loggedHoursPerWeek: [{
    weekStartDate: {
      type: Date,
      required: true
    },
    totalHours: {
      type: Number,
      default: 0
    }
  }],
  status: {
    type: String,
    enum: ['planning', 'active', 'completed', 'canceled', 'on-hold'],
    default: 'planning'
  },
  employeeContributions: [{
    employeeId: {
      type: String,
      required: true
    },
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User'
    },
    role: {
      type: String,
      required: true
    },
    isTeamLead: {
      type: Boolean,
      default: false
    },
    weeklyHours: [{
      weekStartDate: {
        type: Date,
        required: true
      },
      hours: {
        type: Number,
        required: true,
        min: 0,
        max: 168 // maximum hours in a week
      },
      reportedTech: [{
        type: String
      }]
    }]
  }],
  department: {
    type: String,
    required: true,
    index: true
  },
  priority: {
    type: String,
    enum: ['low', 'medium', 'high', 'critical'],
    default: 'medium'
  },
  risks: [{
    description: {
      type: String,
      required: true
    },
    severity: {
      type: String,
      enum: ['low', 'medium', 'high'],
      default: 'medium'
    },
    status: {
      type: String, 
      enum: ['identified', 'mitigated', 'resolved'],
      default: 'identified'
    }
  }]
}, {
  timestamps: true
});

// Create indexes for querying
ProjectSchema.index({ organizationId: 1, createdBy: 1 });
ProjectSchema.index({ organizationId: 1, department: 1 });
ProjectSchema.index({ organizationId: 1, status: 1 });
ProjectSchema.index({ organizationId: 1, "employeeContributions.employeeId": 1 });
ProjectSchema.index({ techStack: 1 });
ProjectSchema.index({ linkedProjects: 1 });

// Pre-save hook to calculate total hours
ProjectSchema.pre<IProject>('save', function(next) {
  // Calculate total hours based on all weekly hours logged by all team members
  let total = 0;
  let weeklyTotals = new Map<string, number>();
  
  this.employeeContributions.forEach(employee => {
    employee.weeklyHours.forEach(week => {
      // Add to total project hours
      total += week.hours;
      
      // Add to weekly totals
      const weekKey = week.weekStartDate.toISOString().split('T')[0];
      const currentTotal = weeklyTotals.get(weekKey) || 0;
      weeklyTotals.set(weekKey, currentTotal + week.hours);
    });
  });
  
  // Update total hours
  this.totalHours = total;
  
  // Update or create loggedHoursPerWeek entries
  this.loggedHoursPerWeek = Array.from(weeklyTotals.entries()).map(([dateStr, hours]) => ({
    weekStartDate: new Date(dateStr),
    totalHours: hours
  }));
  
  next();
});

// Check if model exists before compiling
const Project = mongoose.models.Project as Model<IProject> || 
  mongoose.model<IProject>('Project', ProjectSchema);

export default Project; 

File: models/project.py
from config.database import db
from datetime import datetime

class EmployeeContribution(db.EmbeddedDocument):
    """Embedded document for tracking employee contributions to projects"""
    
    employee_id = db.StringField(required=True)
    role = db.StringField(required=True, choices=['Team Lead', 'Contributor', 'Stakeholder', 'Reviewer'])
    weekly_hours = db.FloatField(default=0.0)
    reported_tech = db.ListField(db.StringField())
    start_date = db.DateTimeField(default=datetime.utcnow)
    end_date = db.DateTimeField()
    active = db.BooleanField(default=True)
    
    def to_dict(self):
        return {
            'employee_id': self.employee_id,
            'role': self.role,
            'weekly_hours': self.weekly_hours,
            'reported_tech': self.reported_tech,
            'start_date': self.start_date,
            'end_date': self.end_date,
            'active': self.active
        }

class Project(db.Document):
    """Project Model representing work contribution data"""
    
    project_id = db.StringField(required=True, unique=True)
    project_title = db.StringField(required=True)
    project_description = db.StringField(required=True)
    tech_stack = db.ListField(db.StringField())
    start_date = db.DateTimeField(required=True)
    end_date = db.DateTimeField()
    department = db.StringField(required=True)
    status = db.StringField(choices=['Planning', 'Active', 'On Hold', 'Completed'], default='Planning')
    priority = db.StringField(choices=['Low', 'Medium', 'High', 'Critical'], default='Medium')
    total_hours = db.FloatField(default=0.0)
    employee_contributions = db.ListField(db.EmbeddedDocumentField(EmployeeContribution))
    
    # Analytics fields
    complexity_score = db.FloatField(default=1.0)  # Calculated based on tech stack, hours, etc.
    impact_score = db.FloatField(default=0.0)      # Business impact rating
    risk_level = db.FloatField(default=0.0)        # Calculated risk assessment
    
    created_at = db.DateTimeField(default=datetime.utcnow)
    updated_at = db.DateTimeField(default=datetime.utcnow)
    
    meta = {
        'collection': 'projects',
        'indexes': [
            'project_id',
            'department',
            'status',
            'start_date',
            'end_date'
        ]
    }
    
    def total_weekly_hours(self):
        """Calculate total weekly hours across all active contributors"""
        return sum(contrib.weekly_hours for contrib in self.employee_contributions if contrib.active)
    
    def active_contributors_count(self):
        """Count active contributors on the project"""
        return sum(1 for contrib in self.employee_contributions if contrib.active)
    
    def to_dict(self):
        """Convert Project document to dictionary"""
        return {
            'project_id': self.project_id,
            'project_title': self.project_title,
            'project_description': self.project_description,
            'tech_stack': self.tech_stack,
            'start_date': self.start_date,
            'end_date': self.end_date,
            'department': self.department,
            'status': self.status,
            'priority': self.priority,
            'total_hours': self.total_hours,
            'employee_contributions': [contrib.to_dict() for contrib in self.employee_contributions],
            'complexity_score': self.complexity_score,
            'impact_score': self.impact_score,
            'risk_level': self.risk_level,
            'created_at': self.created_at,
            'updated_at': self.updated_at
        }
    
    def __str__(self):
        return f"{self.project_title} ({self.status})" 